exec: bad behavior when combining aggregations and projections

I noticed some problematic behavior for queries which project the result of an aggregation. For instance, arithmetic seems to have no effect:
```
root@127.0.0.1:53401/defaultdb> create table t (x int);
CREATE TABLE

Time: 4.454ms

root@127.0.0.1:53401/defaultdb> select count(*)+3 from t;
  ?column?
+----------+
         3
(1 row)

Time: 4.063ms

root@127.0.0.1:53401/defaultdb> set experimental_vectorize=on;
SET

Time: 301µs

root@127.0.0.1:53401/defaultdb> select count(*)+3 from t;
  ?column?
+----------+
         0
(1 row)

Time: 808µs
```
And projecting a comparison causes a panic:
```
root@127.0.0.1:53401/defaultdb> select count(*)>0 from t;
E190709 04:23:51.704829 722 sql/conn_executor.go:713  [n1,client=127.0.0.1:53403,user=root] a SQL panic has occurred while executing "SELECT count(*) > 0 FROM t": interface conversion: coldata.column is []int64, not []bool
E190709 04:23:51.704969 722 util/log/crash_reporting.go:207  [n1,client=127.0.0.1:53403,user=root] a panic has occurred!
panic: interface conversion: coldata.column is []int64, not []bool [recovered]
	panic: panic while executing 1 statements: SELECT count(*) > _ FROM _; caused by interface conversion: coldata.column is []int64, not []bool

goroutine 722 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc001684000, 0x7a9a780, 0xc001614700, 0x6e75f20, 0xc00082b500)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:727 +0x330
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc001684000, 0x7a9a780, 0xc001614700)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:440 +0x61
panic(0x6e75f20, 0xc00082b500)
	/usr/local/Cellar/go/1.12.5/libexec/src/runtime/panic.go:522 +0x1b5
github.com/cockroachdb/cockroach/pkg/sql/exec/coldata.(*memColumn).Bool(0xc000b302c0, 0x0, 0x7afcc00, 0xc000b302c0)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/coldata/vec.go:194 +0x74
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*materializer).Next(0xc000820e00, 0xc000b30280, 0x7a90001, 0xc000c7e290, 0x5a5440e)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/materializer.go:147 +0x8d2
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.Run(0x7a9a840, 0xc00082af90, 0x7aa6380, 0xc000820e00, 0x7a77180, 0xc001118000)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/base.go:171 +0x35
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*ProcessorBase).Run(0xc000820e00, 0x7a9a840, 0xc00082af90)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/processors.go:793 +0x92
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Run(0xc00000cd20, 0x7a9a840, 0xc00082af90, 0x73da150, 0x0, 0x0)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:654 +0x1f2
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0xc000509440, 0xc001178600, 0xc000816090, 0xc000c7e8b0, 0xc001118000, 0xc001684428, 0x0)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:295 +0x35c
github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0xc000509440, 0x7a9a840, 0xc0009c7920, 0xc001684428, 0xc001178600, 0xc000816090, 0x7a9c540, 0xc0013ec360, 0xc001118000)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:859 +0x208
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0xc001684000, 0x7a9a840, 0xc0009c7920, 0xc001684350, 0x3, 0xc16d380, 0xc000b40000, 0xc000c7eb01, 0x0, 0x0, ...)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:902 +0x370
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc001684000, 0x7a9a840, 0xc0009c7920, 0xc001684350, 0xc16d380, 0xc000b40000, 0x0, 0x0)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:734 +0x6e5
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc001684000, 0x7a9a840, 0xc0009c7920, 0x7a9fd40, 0xc0013ce240, 0xc0016752c5, 0x18, 0x0, 0x1, 0x0, ...)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:417 +0xb64
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc001684000, 0x7a9a840, 0xc0009c7920, 0x7a9fd40, 0xc0013ce240, 0xc0016752c5, 0x18, 0x0, 0x1, 0x0, ...)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:99 +0x4ec
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0xc001684000, 0x7a9a780, 0xc0003c3a40, 0x0, 0x0)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1204 +0x1ab0
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc001684000, 0x7a9a780, 0xc001614700, 0xc00024a7c0, 0x5400, 0x15000, 0xc00024a858, 0xc001618560, 0x0, 0x0)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1140 +0x1a3
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc000125200, 0x7a9a780, 0xc001614700, 0xc001684000, 0x5400, 0x15000, 0xc00024a858, 0xc001618560, 0x0, 0x0)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:442 +0xce
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync.func1(0xc0015b15a9, 0xc001698280, 0x7a9a780, 0xc001614700, 0xc001618560, 0xc000125200, 0xc00144d000, 0x7a9e2c0, 0xc001698260, 0xc0015c0480, ...)
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:580 +0x21c
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).processCommandsAsync
	/Users/solon/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:509 +0x17b
```

--- a/pkg/sql/distsqlrun/column_exec_setup.go
+++ b/pkg/sql/distsqlrun/column_exec_setup.go
@@ -142,7 +142,9 @@ func newColOperator(
 			aggSpec.Aggregations[0].FilterColIdx == nil &&
 			aggSpec.Aggregations[0].Func == distsqlpb.AggregatorSpec_COUNT_ROWS &&
 			!aggSpec.Aggregations[0].Distinct {
-			return exec.NewCountOp(inputs[0]), []types.T{types.Int64}, nil
+			op, err = exec.NewCountOp(inputs[0]), nil
+			columnTypes = []semtypes.T{*semtypes.Int}
+			break
 		}
 
 		var groupCols, orderedCols util.FastIntSet

--- a/pkg/sql/exec/execgen/cmd/execgen/projection_ops_gen.go
+++ b/pkg/sql/exec/execgen/cmd/execgen/projection_ops_gen.go
@@ -61,9 +61,9 @@ func (p {{template "opRConstName" .}}) Next(ctx context.Context) coldata.Batch {
 		batch.AppendCol(types.{{.RetTyp}})
 	}
 	vec := batch.ColVec(p.colIdx)
-	col := vec.{{.LTyp}}()[:coldata.BatchSize]
+	col := vec.{{.LTyp}}()
 	projVec := batch.ColVec(p.outputIdx)
-	projCol := projVec.{{.RetTyp}}()[:coldata.BatchSize]
+	projCol := projVec.{{.RetTyp}}()
 	if sel := batch.Selection(); sel != nil {
 		for _, i := range sel {
 			{{(.Assign "projCol[i]" "col[i]" "p.constArg")}}
@@ -105,9 +105,9 @@ func (p {{template "opLConstName" .}}) Next(ctx context.Context) coldata.Batch {
 		batch.AppendCol(types.{{.RetTyp}})
 	}
 	vec := batch.ColVec(p.colIdx)
-	col := vec.{{.RTyp}}()[:coldata.BatchSize]
+	col := vec.{{.RTyp}}()
 	projVec := batch.ColVec(p.outputIdx)
-	projCol := projVec.{{.RetTyp}}()[:coldata.BatchSize]
+	projCol := projVec.{{.RetTyp}}()
 	if sel := batch.Selection(); sel != nil {
 		for _, i := range sel {
 			{{(.Assign "projCol[i]" "p.constArg" "col[i]")}}
@@ -149,11 +149,11 @@ func (p {{template "opName" .}}) Next(ctx context.Context) coldata.Batch {
 		batch.AppendCol(types.{{.RetTyp}})
 	}
 	projVec := batch.ColVec(p.outputIdx)
-	projCol := projVec.{{.RetTyp}}()[:coldata.BatchSize]
+	projCol := projVec.{{.RetTyp}}()
 	vec1 := batch.ColVec(p.col1Idx)
 	vec2 := batch.ColVec(p.col2Idx)
-	col1 := vec1.{{.LTyp}}()[:coldata.BatchSize]
-	col2 := vec2.{{.RTyp}}()[:coldata.BatchSize]
+	col1 := vec1.{{.LTyp}}()
+	col2 := vec2.{{.RTyp}}()
 	if sel := batch.Selection(); sel != nil {
 		for _, i := range sel {
 			{{(.Assign "projCol[i]" "col1[i]" "col2[i]")}}

--- a/pkg/sql/exec/execgen/cmd/execgen/selection_ops_gen.go
+++ b/pkg/sql/exec/execgen/cmd/execgen/selection_ops_gen.go
@@ -105,7 +105,7 @@ func (p *{{template "opConstName" .}}) Next(ctx context.Context) coldata.Batch {
 		}
 
 		vec := batch.ColVec(p.colIdx)
-		col := vec.{{.LTyp}}()[:coldata.BatchSize]
+		col := vec.{{.LTyp}}()
 		var idx uint16
 		n := batch.Length()
 		if vec.MaybeHasNulls() {
@@ -143,8 +143,8 @@ func (p *{{template "opName" .}}) Next(ctx context.Context) coldata.Batch {
 
 		vec1 := batch.ColVec(p.col1Idx)
 		vec2 := batch.ColVec(p.col2Idx)
-		col1 := vec1.{{.LTyp}}()[:coldata.BatchSize]
-		col2 := vec2.{{.RTyp}}()[:coldata.BatchSize]
+		col1 := vec1.{{.LTyp}}()
+		col2 := vec2.{{.RTyp}}()
 		n := batch.Length()
 
 		var idx uint16

--- a/pkg/sql/exec/select_in_tmpl.go
+++ b/pkg/sql/exec/select_in_tmpl.go
@@ -177,7 +177,7 @@ func (si *selectInOp_TYPE) Next(ctx context.Context) coldata.Batch {
 		}
 
 		vec := batch.ColVec(si.colIdx)
-		col := vec._TemplateType()[:coldata.BatchSize]
+		col := vec._TemplateType()
 		var idx uint16
 		n := batch.Length()
 
@@ -247,10 +247,10 @@ func (pi *projectInOp_TYPE) Next(ctx context.Context) coldata.Batch {
 	}
 
 	vec := batch.ColVec(pi.colIdx)
-	col := vec._TemplateType()[:coldata.BatchSize]
+	col := vec._TemplateType()
 
 	projVec := batch.ColVec(pi.outputIdx)
-	projCol := projVec.Bool()[:coldata.BatchSize]
+	projCol := projVec.Bool()
 	projNulls := projVec.Nulls()
 
 	n := batch.Length()

--- a/pkg/sql/logictest/testdata/logic_test/vectorize
+++ b/pkg/sql/logictest/testdata/logic_test/vectorize
@@ -444,3 +444,14 @@ query II
 SELECT * FROM t38753 ORDER BY y;
 ----
 0  NULL
+
+# Regression test for #38752.
+query IIBB
+SELECT count(*), count(*) + 1, count(*) > 4, count(*) + 1 > 4 FROM b
+----
+4  5  false  true
+
+query I
+SELECT * FROM (SELECT count(*) AS x FROM b) WHERE x > 0;
+----
+4

